docs/metadata.rst

```diff
@@ -59,66 +63,190 @@ JSON
+| Format | C Type                   | Python type        | Size in bytes  |
++========+==========================+====================+================+
+| ``x``  | pad byte                 | no value           | as specified   |
++--------+--------------------------+--------------------+----------------+
+| ``c``  | `char`                   | string of length 1 | 1              |
++--------+--------------------------+--------------------+----------------+
+| ``s``  | `char[]`                 | string             | as specified   |
++--------+--------------------------+--------------------+----------------+
+| ``p``  | `char[]`                 | string             | as specified   |
++--------+--------------------------+--------------------+----------------+
+
+For the ``'s'`` format character, the number prefixed is interpreted as the length in
+bytes, for example,
+``'10s'`` means a single 10-byte string. For packing, the string is
+truncated or padded with null bytes as appropriate to make it fit. For
+unpacking, the resulting bytes object always has exactly the specified number
+of bytes, unless ``nullTerminated`` is ``true``, in which case it ends at the first
+``null``. As a special case, ``'0s'`` means a single, empty string.
+
+The ``'p'`` format character encodes a "Pascal string", meaning a short
+variable-length string stored in a fixed number of bytes, given by the count.
+The first byte stored is the length of the string, or 255, whichever is
+smaller.  The bytes of the string follow.  If the string to encode is too long
+(longer than the count minus 1), only the leading
+``count-1`` bytes of the string are stored.  If the string is shorter than
+``count-1``, it is padded with null bytes so that exactly count bytes in all
+are used.  Note that strings specified with this format cannot be longer than 255.
+
+Strings that are longer than the specified length will be silently truncated,
+note that the length is in bytes, not characters.
+
+The string encoding can be set with ``stringEncoding`` which defaults to ``utf-8``.
+A list of possible encodings is `here <https://docs.python.org/3.7/library/codecs.html#standard-encodings>`_
+
+For most cases, where there are no ``null`` characters in the metadata
+``{"type":"string", "binaryFormat": "1024s", "nullTerminated": True}`` is a good option
+with the size set to that appropriate for the strings to be encoded.
+
+~~~~~~~~~~~~~
+Padding bytes
+~~~~~~~~~~~~~
+
+Unused padding bytes (for compatibility) can be added with a schema entry like::
+
+    {"type": "null", "binaryFormat":"5x"} # 5 padding bytes
+
+~~~~~~
+Arrays
+~~~~~~
+
+The codec stores the length of the array before the array data. The format used for the
+length of the array can be chosen with ``arrayLengthFormat`` which must be one
+of ``B``, ``H``, ``I``, ``L`` or ``Q`` which have the same meaning as in the numeric
+types above. ``Q`` is the default. As an example::
```

Comment:
see #769

Reply:
Fixed in https://github.com/tskit-dev/tskit/pull/741/commits/fecbe450ca4fa8903deef1b82790074d824947b4 with added error message on overflow.